File: src/algorithms/hash-maps/tracking/find-all-duplicates/sources/FindAllDuplicates_test.cpp
#include "FindAllDuplicates.cpp"
#include <cassert>
#include <iostream>
#include <vector>

int main() {
    assert((findAllDuplicates({4, 3, 2, 7, 8, 2, 3, 1}) == std::vector<int>{2, 3}));
    assert((findAllDuplicates({1, 1, 2}) == std::vector<int>{1}));
    assert(findAllDuplicates({1, 2, 3}).empty());
    assert(findAllDuplicates({}).empty());
    assert((findAllDuplicates({5, 5}) == std::vector<int>{5}));
    assert((findAllDuplicates({1, 2, 1, 2}) == std::vector<int>{1, 2}));
    assert(findAllDuplicates({7}).empty());
    assert((findAllDuplicates({3, 3, 3}) == std::vector<int>{3, 3}));

    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: src/algorithms/hash-maps/tracking/find-all-duplicates/__tests__/FindAllDuplicates_test.java
import java.util.*;

public class FindAllDuplicates_test {
    public static void main(String[] args) {
        assert FindAllDuplicates.findAllDuplicates(new int[]{4, 3, 2, 7, 8, 2, 3, 1}).equals(Arrays.asList(2, 3));
        assert FindAllDuplicates.findAllDuplicates(new int[]{1, 1, 2}).equals(Arrays.asList(1));
        assert FindAllDuplicates.findAllDuplicates(new int[]{1, 2, 3}).equals(Collections.emptyList());
        assert FindAllDuplicates.findAllDuplicates(new int[]{}).equals(Collections.emptyList());
        assert FindAllDuplicates.findAllDuplicates(new int[]{5, 5}).equals(Arrays.asList(5));
        assert FindAllDuplicates.findAllDuplicates(new int[]{1, 2, 1, 2}).equals(Arrays.asList(1, 2));
        assert FindAllDuplicates.findAllDuplicates(new int[]{7}).equals(Collections.emptyList());
        assert FindAllDuplicates.findAllDuplicates(new int[]{3, 3, 3}).equals(Arrays.asList(3, 3));

        System.out.println("All tests passed!");
    }
}
